packages/backend-api/install/tests/ShopBundle/Smoke/BackendApiCreateProductTest.php

```diff
@@ -0,0 +1,105 @@
+ */
+class BackendApiCreateProductTest extends OauthTestCase
+{
+    public function testCreatePostProduct(): void
+    {
+        $product = [
+            'name' => [
+                'en' => 'X tech',
+                'cs' => 'název',
+            ],
+            'hidden' => true,
+            'sellingDenied' => true,
+            'sellingFrom' => '2019-07-16T00:00:00+00:00',
+            'sellingTo' => '2020-07-16T00:00:00+00:00',
+            'catnum' => '123456 co',
+            'ean' => 'E12346B',
+            'partno' => 'P123456',
+            'shortDescription' => [
+                1 => '<b>desc',
+                2 => '<b>popisek',
+            ],
+            'longDescription' => [
+                1 => '<b>desc long',
+                2 => '<b>popisek dlouhy',
+            ],
+        ];
+        $response = $this->runOauthRequest('POST', '/api/v1/products', $product);
+
+        $this->assertSame(201, $response->getStatusCode());
+
+        $location = $response->headers->get('Location');
+        $this->assertProductLocation($location);
+
+        $uuid = $this->extractUuid($location);
+
+        $foundProduct = $this->getProduct($uuid);
+        $this->assertSame($product, $foundProduct);
+    }
+
+    /**
+     * @param string $location
+     */
+    private function assertProductLocation(string $location): void
+    {
+        $expectedLocationStart = sprintf('%s/api/v1/products/', $this->getDomainBaseUrl());
+        $this->assertStringStartsWith($expectedLocationStart, $location);
+    }
+
+    /**
+     * @param string $uuid
+     * @return array
+     */
+    private function getProduct(string $uuid): array
+    {
+        $response = $this->runOauthRequest('GET', '/api/v1/products/' . $uuid);
+        $this->assertSame(200, $response->getStatusCode());
+        $data = json_decode($response->getContent(), true);
+        unset($data['uuid']);
+        return $data;
+    }
+
+    public function testValidationError(): void
+    {
+        $product = [
+            'name' => [
+                'en' => 'X tech',
+                'xx' => 'název',
+            ],
+            'hidden' => false,
+            'sellingFrom' => null,
+            'sellingTo' => '12.1.2019',
+            'ean' => 'E12346B',
+            'partno' => 'P123456',
+            'shortDescription' => [
+                1 => '<b>desc',
+                2 => '<b>popisek',
+            ],
+            'longDescription' => [
+                1 => '<b>desc long',
+                2 => '<b>popisek dlouhy',
+            ],
+        ];
+        $response = $this->runOauthRequest('POST', '/api/v1/products', $product);
+
+        $this->assertSame(400, $response->getStatusCode());
+
+        $expectedResponse = [
+            'code' => 400,
+            'message' => 'During validation of data has been found these errors: [name][xx] This field was not expected., [sellingTo] The value "12.1.2019" is not a valid DateTime::ATOM format.',
```

Comment:
structure of errors should be different, see https://github.com/shopsys/shopsys/blob/f4a86c067fcc77f3cc057a13122e9803b19299fb/docs/backend-api/api-methods.md#add-product "error response"

Reply:
So we agreed that `ProductApiDataValidatorInterface::validate` will return an array of errors. And controller will always check error count and if it is `> 0` then it will return a different `View`